### PR TITLE
Feature/refactor object pool

### PR DESCRIPTION
**Pull request by gnusi**
Remove custom `std::unique_ptr`-like smart pointers in object pools.

### CODE REVIEW

File: core/utils/object_pool.hpp [hunk 1]
```diff
@@ -115,79 +95,61 @@ class concurrent_stack : private util::noncopyable {
...
 
-  value_type& value() noexcept {
-    return value_;
-  }
+  value_type& value() noexcept { return value_; }
 
-  auto lock_read() {
-    return make_shared_lock(lock_);
-  }
+  auto lock_read() { return make_shared_lock(lock_); }
```
Comment (MBkkt) [inline core/utils/object_pool.hpp:137]: std::shared_lock{lock_}

File: core/utils/object_pool.hpp [hunk 2]
```diff
@@ -115,79 +95,61 @@ class concurrent_stack : private util::noncopyable {
...
 
-  auto lock_write() {
-    return make_unique_lock(lock_);
-  }
+  auto lock_write() { return make_unique_lock(lock_); }
```
Comment (MBkkt) [inline core/utils/object_pool.hpp:139]: std::unique_lock{lock_}

File: core/utils/object_pool.hpp [hunk 3]
```diff
@@ -253,19 +200,16 @@ class bounded_object_pool {
...
       return std::shared_ptr<element_type>(
-        ptr,
-        [slot] (element_type*) mutable noexcept {
-          reset_impl(slot);
-      });
+          ptr, [slot](element_type*) mutable noexcept { reset_impl(slot); });
     }
 
-    operator bool() const noexcept { return &EMPTY_SLOT != slot_; }
+    operator bool() const noexcept { return nullptr != slot_; }
```
Comment (MBkkt) [inline core/utils/object_pool.hpp:212]: Try explicit?

File: core/utils/object_pool.hpp [hunk 4]
```diff
@@ -115,79 +95,61 @@ class concurrent_stack : private util::noncopyable {
...
   // CMPXCHG16B requires that the destination
   // (memory) operand be 16-byte aligned
   struct alignas(IRESEARCH_CMPXCHG16B_ALIGNMENT) concurrent_node {
     explicit concurrent_node(node_type* node = nullptr) noexcept
-      : version{0}, node{node} {
-    }
+        : version{0}, node{node} {}
 
-    uintptr_t version; // avoid aba problem
+    uintptr_t version;  // avoid aba problem
```
Comment (MBkkt) [inline core/utils/object_pool.hpp:109]: Btw I thought about ifdef that remove version for arm

File: core/utils/object_pool.hpp [hunk 5]
```diff
@@ -253,19 +200,16 @@ class bounded_object_pool {
     std::shared_ptr<element_type> release() {
```
Comment (MBkkt) [inline core/utils/object_pool.hpp:200]: So this release necessary? I mean ptr semantic still move only

**Review by MBkkt — APPROVED** [review]

**Review by Dronplane — APPROVED** [review]
LGTM

**Review by Dronplane — APPROVED** [review]

**Pull request merged**